src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -277,6 +277,20 @@ impl ScalarFunc {
         }
         Ok(Some(result))
     }
+
+    #[inline]
+    pub fn date_diff(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let l_t: Cow<Time> = try_opt!(self.children[0].eval_time(ctx, row));
+        if l_t.is_zero() {
```

Comment:
I think we should use `invalid_zero()` here. @breeswish

Reply:
Do you mean this function?
https://github.com/tikv/tikv/blob/670a95aadc5ba90cf45b188da639cf7299bf4942/src/coprocessor/codec/mysql/time/mod.rs#L264

I followed convention in the current code, just curious
1) why not extract out a function for DateTime validation. I saw the validation everywhere in the file and can cause inconsistency issue
2) Why zero DateTime is not allowed

Reply:
Maybe @breeswish or @zz-jason can help to answer this.

Reply:
@huachaohuang @breeswish @zz-jason 

DateDiff fixed by always setting time to be zero:
        let lhs_date_time = lhs.get_time().with_hour(0).unwrap()
            .with_minute(0).unwrap().with_second(0).unwrap().with_nanosecond(0).unwrap();

I can submit another CR to refactor DateTime validation. Let me know if you have other comments. 
Gentle reminder, the two questions I raised are still open.

Reply:
Hi, @edwardpku 
> Why zero DateTime is not allowed

MySQL permits you to store dates where the day or month and day are zero in a DATE or DATETIME column. This is useful for applications that need to store birthdates for which you may not know the exact date. In this case, you simply store the date as '2009-00-00' or '2009-01-00'. If you store dates such as these, you should not expect to get correct results for functions such as DATE_SUB() or DATE_ADD() that require complete dates. 

And we check `invalid_zero` here to be compatiable with the behavior of `DATEDIFF` in MySQL.

> why not extract out a function for DateTime validation. I saw the validation everywhere in the file and can cause inconsistency issue

MySQL does not hold a unified validation for Datetime for different situation. For example, different SQL_MODE may cause behavior during calculation for a same query.

Reply:
@XuHuaiyu  
Thank you so much for your reply. I made anther revision, please take a look